File: BOJ/7569.java
import java.lang.*;
import java.util.*;

public class Main {
    public static void main(String[] args) {
        class pos {
            int h, r, c;

            public pos(int h_, int r_, int c_) {
                h = h_;
                r = r_;
                c = c_;
            }
        }

        Scanner sc = new Scanner(System.in);
        int C = sc.nextInt();
        int R = sc.nextInt();
        int H = sc.nextInt();
        int[][][] box = new int[H][R][C];
        int[][][] v = new int[H][R][C];
        Queue<pos> q = new LinkedList<>();

        for (int h = 0; h < H; h++) {
            for (int r = 0; r < R; r++) {
                for (int c = 0; c < C; c++) {
                    box[h][r][c] = sc.nextInt();
                    if (box[h][r][c] == 1)
                        q.add(new pos(h, r, c));
                }
            }
        }

        int[] dh = {1, -1, 0, 0, 0, 0};
        int[] dr = {0, 0, 1, -1, 0, 0};
        int[] dc = {0, 0, 0, 0, 1, -1};

        while (!q.isEmpty()) {
            pos tmp = q.poll();
            int h = tmp.h;
            int r = tmp.r;
            int c = tmp.c;
            for (int i = 0; i < 6; i++) {
                int nh = h + dh[i];
                int nr = r + dr[i];
                int nc = c + dc[i];
                if (nh < 0 || nr < 0 || nc < 0 || nh >= H || nr >= R || nc >= C || v[nh][nr][nc] == 1 || box[nh][nr][nc] != 0)
                    continue;
                box[nh][nr][nc] = box[h][r][c] + 1;
                v[nh][nr][nc] = 1;
                q.add(new pos(nh, nr, nc));
            }
        }

        int ans = 0;
        for (int[][] floor : box) {
            for (int[] row : floor) {
                for (int e : row) {
                    if (e == 0) {
                        System.out.println(-1);
                        System.exit(0);
                    } else {
                        ans = Math.max(ans, e);
                    }
                }
            }
        }
        System.out.println(ans - 1);
    }
}

File: BOJ/7569.py
from collections import deque

M, N, H = map(int, input().split())

box = []
for _ in range(H):
    board = [list(map(int, input().split())) for k in range(N)]
    box.append(board)


q = deque()

for h, board in enumerate(box):
    for r, row in enumerate(board):
        for c, e in enumerate(row):
            if e == 1:
                q.append((h, r, c))

isVisited = [[[-1 for _ in range(M)] for _ in range(N)] for _ in range(H)]

dr = [1, -1, 0, 0, 0, 0]
dc = [0, 0, 1, -1, 0, 0]
dh = [0, 0, 0, 0, 1, -1]

while (q):
    h, r, c = q.popleft()
    for i in range(6):
        nh, nr, nc = h + dh[i], r + dr[i], c + dc[i]
        if nh < 0 or nr < 0 or nc < 0 or nh >= H or nr >= N or nc >= M or isVisited[nh][nr][nc] == 1 or box[nh][nr][nc] != 0:
            continue
        box[nh][nr][nc] = box[h][r][c] + 1
        isVisited[nh][nr][nc] = 1
        q.append((nh, nr, nc))

ans = 0
for floor in box:
    for row in floor:
        for e in row:
            if e == 0:
                print(-1)
                exit()
            else:
                ans = max(ans, e)
print(ans-1)
